textile/features.textile

```diff
@@ -1499,6 +1500,29 @@ h4. ChannelMetrics
 ** @(CHM2e)@ @publishers@ integer - the number of realtime attachments which are able to publish messages to the channel (that is, they have the `publish` capability and have not specified a @ChannelMode@ that excludes @PUBLISH@)
 ** @(CHM2f)@ @subscribers@ integer - the number of realtime attachments which are receiving messages on the channel (that is, they have the `subscribe` capability and have not specified a @ChannelMode@ that excludes @SUBSCRIBE@)
 
+h4. BatchResult
+
+* @(BAR1)@ Contains information about the results of a batch operation
+** @(BAR2)@ The attributes of @BatchResult@ consist of:
+** @(BAR2a)@ @successCount@ number - the number of successful operations
+** @(BAR2b)@ @failureCount@ number - the number of unsuccessful operations
+** @(BAR2c)@ @results@ array - an array of results for the batch operation
+
+h4. BatchSpec
+
+* @(BSP1)@ Describes the messages that should be published by a batch publish operation, and the channels to which they should be published
+* @(BSP2)@ The attributes of @BatchSpec@ consist of:
+** @(BSP2a)@ @channels@ an array of strings - the names of the channels to which all of the messages contained in the @messages@ attribute should be published
+** @(BSP2b)@ @messages@ an array of @Message@ objects - the names of the channels to which all of the messages contained in the @messages@ attribute should be published
+
+h4. BatchPublishResult
+
+* @(BPR1)@ Contains information about the result of publishes requested by a single @BatchSpec@ for a single channel
+* @(BPR2)@ The attributes of @BatchPublishChannelResult@ consist of:
+** @(BPR2a)@ @channel@ string - the name of the channel
+** @(BPR2b)@ @messageId@ string - an optional string containing the @messageId@ prefix for the published message(s), if successful
```

Comment:
If you publish multiple messages to the same channel in a batch, the messages will all start with the same prefix,  IIRC the message IDs look like `<some_prefix>:n`, where `n` is a unique non-negative integer. The REST API returns these in order for clients to identify which messages were published as part of a batch request

Reply:
A couple of questions, in that case:

- isn't the attribute name `messageId` a bit misleading, if it's actually just a prefix of the message’s ID?
- if I've understood the structure of the result of `batchPublish(BatchPublishSpec)` correctly (which I may not have done — see my question elsewhere in this PR), then we have one `BatchPublishSuccessResult` for each message that resulted from the batch operation, so couldn't we actually include the full message ID instead of just the prefix?

Reply:
When you do a normal realtime/rest publish of an array of Messages, they're handled atomically. the client library bundles them up into a single ProtocolMessage, that either succeeds or fails. That ProtocolMessage is the thing that actually has the real messageId. But on the subscribe side, because the user doesn't actually interact with the ProtocolMessage only the Messages but we still want to expose that information somehow, the receiving sdk takes the protocolmessage message id (eg `abc123:0`), appends an index for each of the Messages inside, and sets it as the Message.id (eg `abc123:0:0` through `abc123:0:3`). See [TM2a](https://sdk.ably.com/builds/ably/specification/main/features/#TM2a).

In basic pubsub this is fairly transparent to the user; the realtime system only knows about the 'real' messageId, the user only knows the fake ones. But the distinction is kinda important to the guarantees of what the system does (eg the whole ProtocolMessage will succeed or be rejected for a rate limit atomically), and so it is documented. Eg if the user wants to do idempotent publishing with customer-specified ids, [they will need to learn about this](https://faqs.ably.com/client-specified-message-id-restrictions-for-multiple-messages-published-atomically), because the idempotency applies to the ProtocolMessage, not the individual Message. And in the batch api, this abstraction is exposed too -- all messages in a single batchspec are published atomically (for each channel in that batchspec), so by specifying multiple batchspecs, the user has control over which groups of messages are published atomically. The return type reflects this by having one BatchPublishResult for each group of messages atomically published to the system.

It's a bit of a balance between making it easy for newbies and allowing advanced users more insight into and control over what's actually happening. Ofc we may have gotten that balance wrong here. There's still time to propose a change to the v3 batch response type if you want to do so, no customers are using it yet. (eg plausibly realtime could anticipate the numbering that the subscribing sdks will give the individual messages and return `messageIds: ["foo:0:0", "foo:0:1"]` instead of `messageId: "foo"`)

Reply:
Eh, maybe.

Part of my reluctance is that `BatchPublishSuccessResult` isn't actually anything specific to the batch api. It's literally just the response body you get from a normal REST publish (try it). It'd be wrong from an api coherence point of view to changing the name of the property for the batch publish api but not the normal publish api.

So the question then becomes, should we change it in both for protocol v3? And... maybe, but I'm not quite convinced it's worth the faff of a breaking change, even if yeah it would be more correct (given that, afaik, no customers have ever complained about this. almost no-one ever even looks at the message id of a publish afaik, most people only care if it succeeded or failed)